addons/pos_coupon/models/coupon_program.py

```diff
@@ -0,0 +1,85 @@
+# -*- coding: utf-8 -*-
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+
+# NOTE Use black to automatically format this code.
+
+from odoo import api, fields, models, _
+
+import ast
+
+class CouponProgram(models.Model):
+    _inherit = "coupon.program"
+
+    pos_config_ids = fields.Many2many(
+        "pos.config",
+        relation="program_pos_config_rel",
+        string="Point of Sales",
+        readonly=True,
+    )
+    pos_order_line_ids = fields.One2many(
+        "pos.order.line",
+        "program_id",
+        string="PoS Order Lines",
+        help="Order lines where this program is applied.",
+    )
+    promo_barcode = fields.Char(
+        "Barcode",
+        default=lambda self: self.env["coupon.coupon"]._generate_code(),
+        help="A technical field used as an alternative to the promo_code. "
+        "This is automatically generated when promo_code is changed.",
+    )
+    pos_order_ids = fields.Many2many(
+        "pos.order", help="The PoS orders where this program is applied.",
+    )
+    pos_order_count = fields.Integer(
+        "PoS Order Count", compute="_compute_pos_order_count"
+    )
+    valid_product_ids = fields.Many2many(
+        "product.product",
+        "Valid Products",
+        compute="_compute_valid_product_ids",
+        help="These are the products that are valid in this program.",
+    )
+    valid_partner_ids = fields.Many2many(
+        "res.partner",
+        "Valid Partners",
+        compute="_compute_valid_partner_ids",
+        help="These are the partners that can avail this program.",
+    )
+
+    @api.depends("pos_order_ids")
+    def _compute_pos_order_count(self):
+        for program in self:
+            program.pos_order_count = len(program.pos_order_ids)
+
+    def write(self, vals):
+        if "promo_code" in vals:
+            vals.update({"promo_barcode": self.env["coupon.coupon"]._generate_code()})
+        return super(CouponProgram, self).write(vals)
+
+    def action_view_pos_orders(self):
+        self.ensure_one()
+        return {
+            "name": _("PoS Orders"),
+            "view_mode": "tree,form",
+            "res_model": "pos.order",
+            "type": "ir.actions.act_window",
+            "domain": [("id", "in", self.pos_order_ids.ids)],
+            "context": dict(self._context, create=False),
+        }
+
+    @api.depends("rule_products_domain")
+    def _compute_valid_product_ids(self):
```

Comment:
shouldn't we put those two in coupon as they have the same purpose as get_valid_products and is_valid partners?

Reply:
You are right, the two have overlap with the ones in sale_coupon. I'll check if this can be generalized in coupon.

Reply:
Our use case of `rule_products_domain` and `rule_partners_domain` has difference in the sale_coupon's use case. In sale_coupon, to validate products/partner, they reconstruct the `rule_%s_domain` by adding new condition based on given products/partners, then they search product/partner using the new domain. (I say this is efficient.) They don't need all the valid products/partners all the time because they can always perform a search in the orm.

However, in pos_coupon's case, we are loading all the valid products and partners, then we save them in Set for rule validation in the order.

So I guess the valid_product_ids and valid_partner_ids fields are not useful in sale_coupon.

Do you think we should do the same strategy as the sale_coupon when validating products/partner in the order? In my opinion, what we are doing now is better because we don't suffer from network inefficiency even though validating products/partner using _get_valid_products/_is_valid_partner is more efficient in the orm.

Reply:
ok, no problem